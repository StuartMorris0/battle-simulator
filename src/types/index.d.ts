declare interface PlayerHealth {
  current: number;
  lost: number | undefined;
}

declare interface Player {
  playerName: string;
  dice1: number;
  dice2: number;
  health: PlayerHealth;
  hasWon: boolean;
}
