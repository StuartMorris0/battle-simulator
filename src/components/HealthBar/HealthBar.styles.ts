import { motion } from 'framer-motion';
import styled from 'styled-components';

import { IHealthBarProps } from './HealthBar.types';

export const Container = styled.div`
  display: flex;
  height: 120px;
  flex-direction: column;
  position: relative;

  @media (min-width: 769px) {
    height: 260px;
  }
`;

export const LostHealth = styled(motion.div)`
  position: absolute;
  top: 50%;
  left: 25%;
  background: white;
  border-radius: 26px;
  padding: 10px;
  border: 1px solid #f1f1f1;
  font-size: 20px;
  font-weight: bold;
`;

export const Score = styled.div`
  height: 2rem;
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;
  font-weight: bold;
`;

export const HealthWrapper = styled.div`
  background-color: white;
  flex-grow: 1;
  border-radius: 3px;
  width: 32px;
  display: flex;
  align-items: flex-end;
  padding: 2px;
`;

export const CurrentHealth = styled.div<IHealthBarProps>`
  background-color: #51da51;
  border-radius: 4px;
  display: block;
  width: 100%;
  height: ${({ current }) => current}%;
  transition: height 1s ease-in-out;

  background-color: ${({ current }) => current < 40 && `#ffc800`};
  background-color: ${({ current }) => current < 20 && `#fb4b17`};
`;
