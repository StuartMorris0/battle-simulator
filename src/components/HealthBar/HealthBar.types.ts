export declare interface IHealthBarProps {
  current: number;
  lost: number | undefined;
}
