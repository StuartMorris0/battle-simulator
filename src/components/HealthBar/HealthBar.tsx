import React from 'react';

import { Container, CurrentHealth, HealthWrapper, LostHealth, Score } from './HealthBar.styles';
import { IHealthBarProps } from './HealthBar.types';

const Bar: React.FC<IHealthBarProps> = ({ current, lost }: IHealthBarProps) => (
  <>
    <Container data-testid="Bar">
      {lost && (
        <LostHealth
          initial={{ opacity: 0, scale: 1 }}
          animate={{
            scale: [1.1, 1, 1.1, 1],
            x: 25,
            y: -25,
            opacity: [1, 1, 0.75, 0],
          }}
          transition={{
            duration: 2.5,
            ease: 'easeInOut',
            times: [0, 0.25, 0.5, 0.75, 1],
          }}
          exit={{
            opacity: 0,
            scale: 0,
          }}
        >
          {`-${lost}`}
        </LostHealth>
      )}
      <HealthWrapper>
        <CurrentHealth lost={lost} current={current > 0 ? current : 0} />
      </HealthWrapper>
      <Score>{current > 0 ? current : 0}</Score>
    </Container>
  </>
);

export default Bar;
