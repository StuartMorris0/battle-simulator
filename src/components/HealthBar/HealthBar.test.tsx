import React from 'react';

import { render } from 'global/utils/test-utils';

import HealthBar from '.';
import { IHealthBarProps } from './HealthBar.types';

describe('HealthBar Component', () => {
  let props: IHealthBarProps;
  beforeEach(() => {
    props = {
      current: 100,
      lost: undefined,
    };
  });

  const renderComponent = () => render(<HealthBar {...props} />);

  test('renders', () => {
    const { getByTestId, getByText } = renderComponent();
    expect(getByTestId('HealthBar')).toBeInTheDocument();
    expect(getByText('100')).toBeInTheDocument();
  });

  test('renders lost health', () => {
    props.lost = 6;
    const { getByText } = renderComponent();
    expect(getByText('-6')).toBeInTheDocument();
  });

  test('renders random health', () => {
    props.current = 25;
    const { getByTestId, getByText } = renderComponent();
    expect(getByTestId('HealthBar')).toBeInTheDocument();
    expect(getByText('25')).toBeInTheDocument();
  });

  test('renders 0 for negative health', () => {
    props.current = -25;
    const { getByTestId, getByText } = renderComponent();
    expect(getByTestId('HealthBar')).toBeInTheDocument();
    expect(getByText('0')).toBeInTheDocument();
  });

  test('render snaphost', () => {
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });
});
