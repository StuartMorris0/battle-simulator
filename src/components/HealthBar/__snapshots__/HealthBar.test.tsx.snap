// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthBar Component render snaphost 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 120px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
}

.c3 {
  height: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: white;
  font-weight: bold;
}

.c1 {
  background-color: white;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  border-radius: 3px;
  width: 32px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  padding: 2px;
}

.c2 {
  background-color: #51da51;
  border-radius: 4px;
  display: block;
  width: 100%;
  height: 100%;
  -webkit-transition: height 1s ease-in-out;
  transition: height 1s ease-in-out;
}

@media (min-width:769px) {
  .c0 {
    height: 260px;
  }
}

<div
  class="c0"
  data-testid="HealthBar"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    />
  </div>
  <div
    class="c3"
  >
    100
  </div>
</div>
`;
